js/modules/k6/http/async_request_test.go

```diff
@@ -0,0 +1,290 @@
+		t.Parallel()
+		ts := newTestCase(t)
+
+		sr := ts.tb.Replacer.Replace
+		err := runOnEventLoop(ts.runtime, sr(`
+				var reqUrl = "HTTPBIN_URL/cookies"
+				var res = await http.asyncRequest("GET", reqUrl);
+				var jar = new http.CookieJar();
+
+				jar.set("HTTPBIN_URL/cookies", "key", "value");
+				res = await http.asyncRequest("GET", reqUrl, null, { cookies: { key2: "value2" }, jar: jar });
+
+				if (res.json().key != "value") { throw new Error("wrong cookie value: " + res.json().key); }
+
+				if (res.status != 200) { throw new Error("wrong status: " + res.status); }
+				if (res.request["method"] !== "GET") { throw new Error("http request method was not \"GET\": " + JSON.stringify(res.request)) }
+				if (res.request["body"].length != 0) { throw new Error("http request body was not null: " + JSON.stringify(res.request["body"])) }
+				if (res.request["url"] != reqUrl) {
+					throw new Error("wrong http request url: " + JSON.stringify(res.request))
+				}
+				if (res.request["cookies"]["key2"][0].name != "key2") { throw new Error("wrong http request cookies: " + JSON.stringify(JSON.stringify(res.request["cookies"]["key2"]))) }
+				if (res.request["headers"]["User-Agent"][0] != "TestUserAgent") { throw new Error("wrong http request headers: " + JSON.stringify(res.request)) }
+				`))
+		assert.NoError(t, err)
+	})
+	t.Run("NonEmptyBody", func(t *testing.T) {
+		t.Parallel()
+		ts := newTestCase(t)
+
+		sr := ts.tb.Replacer.Replace
+		err := runOnEventLoop(ts.runtime, sr(`
+				var res = await http.asyncRequest("POST", "HTTPBIN_URL/post", {a: "a", b: 2}, {headers: {"Content-Type": "application/x-www-form-urlencoded; charset=utf-8"}});
+				if (res.status != 200) { throw new Error("wrong status: " + res.status); }
+				if (res.request["body"] != "a=a&b=2") { throw new Error("http request body was not set properly: " + JSON.stringify(res.request))}
+				`))
+		assert.NoError(t, err)
+	})
+	t.Run("Concurrent", func(t *testing.T) {
+		t.Parallel()
+		ts := newTestCase(t)
+		sr := ts.tb.Replacer.Replace
+		err := runOnEventLoop(ts.runtime, sr(`
+            let start = Date.now()
+            let p1 = http.asyncRequest("GET", "HTTPBIN_URL/delay/200ms").then(() => { return Date.now() - start})
+            let p2 = http.asyncRequest("GET", "HTTPBIN_URL/delay/100ms").then(() =>  { return Date.now() - start})
+            let time1 = await p1;
+            let time2 = await p2;
+            if (time1 < time2) {
+                throw("request that should've taken 200ms took less time then one that should take 100ms " + time1 +">" + time2 )
+            }
+
+				`))
+		assert.NoError(t, err)
+	})
+}
+
+func TestAsyncRequestResponseCallbackRace(t *testing.T) {
+	// This test is here only to tease out race conditions
+	t.Parallel()
+	ts := newTestCase(t)
+	err := ts.runtime.VU.Runtime().Set("q", func(f func()) {
+		rg := ts.runtime.EventLoop.RegisterCallback()
+		time.AfterFunc(time.Millisecond*5, func() {
```

Comment:
This is a very confusing test, and given the very small duration here and the unreliable state of CI, it might lead to a lot of false positives, and be more trouble than it's worth.

Have you run into these race conditions locally and fixed them, and do you really think this test would be helpful?

Reply:
Well, sure, but hopefully a test would replicate a scenario that could possibly fail, or that failed previously and was fixed, and not just model a hypothetical scenario that has little to no chance of happening in real world usage.

But if this failed at some point for you before and now it's fixed, then it's good to keep around. And if it ends up being flaky in the long run, we can consider removing or fixing it then.